src/Microsoft.AspNetCore.Http/StreamPipeWriter.cs

```diff
@@ -0,0 +1,299 @@
+        }
+
+        /// <inheritdoc />
+        public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
+        {
+            return FlushAsyncInternal(cancellationToken);
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncInternal(CancellationToken cancellationToken = default)
+        {
+            if (_bytesWritten == 0)
+            {
+                return new FlushResult(isCanceled: false, IsCompletedOrThrow());
+            }
+
+            // Wrap the provided cancellationToken with an internal one
+            // to allow CancelPendingFlush to cancel writes and flushes
+            if (InternalTokenSource.IsCancellationRequested)
+            {
+                // If CancelPendingFlush was already called, we need to return a FlushResult with
+                // Canceled. At this point, we create a fresh CTS that isn't canceled.
+                // PERF: we only create a new CTS if the previous one wasn't canceled.
+                // Otherwise, we would need to create a new CTS for every call to FlushAsync.
+                var result = new FlushResult(isCanceled: true, IsCompletedOrThrow());
+                InternalTokenSource = new CancellationTokenSource();
+                return result;
+            }
+
+            var token = InternalTokenSource.Token;
+            if (cancellationToken != CancellationToken.None)
+            {
+                // PERF: only link token sources if FlushAsync was provided a cancellation token.
+                token = CancellationTokenSource.CreateLinkedTokenSource(
+                    cancellationToken,
+                    InternalTokenSource.Token).Token;
+            }
+
+            try
+            {
+                // Write all completed segments and whatever remains in the current segment
+                // and flush the result.
+                if (_completedSegments != null && _completedSegments.Count > 0)
+                {
+                    // We need to treat the first completed segment uniquely because we allow 
+                    // partial consumption of the first CompletedSegment.
+                    // Example:
+                    // A user calls GetMemory(11) to move the _currentSegment _position to 11 and then call FlushAsync().
+                    // Afterwards, they call GetMemory(1000), which would return Memory in _currentSegment.
+                    // Afterwards, they call GetMemory(8000), which would return a new segment, which moves _currentSegment to _completedSegments
+                    // Then the user calls FlushAsync().
+                    // At this point, the firstSegment needs to be treated differently because it was only partially written to.
+                    // To fix this, we use _previousSegmentPosition to store how far we were into the first _completedSegment.
+                    var firstSegment = _completedSegments[0];
```

Comment:
@davidfowl @pakrym this logic is pretty complicated due to supporting partial writes in a CompletedSegment (rather than resetting every time). I feel a linked list may suit us better here.